### PR TITLE
[compiler] Remove ConstAPI struct

### PR DESCRIPTION
**Pull request by staticfloat**
`ConstAPI` predates the effect analysis framework, and was relying on an
ad-hoc implementation of effect analysis that only worked for functions
of less than 15 statements.  As a result, it wasn't used much anymore,
but it could cause trouble with external `AbstractInterpreters` that may
want to cache additional code metadata within `CodeInstance`s.

This commit simply removes this, expecting that the existing effect
analysis will pick up the slack.

### CODE REVIEW

**Review by Keno — APPROVED** [review]

File: base/compiler/typeinfer.jl [hunk 1]
```diff
@@ -300,9 +292,10 @@ function CodeInstance(
     local const_flags::Int32
     result_type = result.result
     @assert !(result_type isa LimitedAccuracy)
-    if inferred_result isa ConstAPI
+
+    if isa(result_type, Const) && is_total(result.ipo_effects)
```
Comment (aviatesk) [inline base/compiler/typeinfer.jl:296]: I think we only require `is_foldable && is_nothrow` here?

File: base/compiler/typeinfer.jl [hunk 2]
```diff
@@ -379,7 +372,7 @@ function transform_result_for_cache(interp::AbstractInterpreter,
         inferred_result = maybe_compress_codeinfo(interp, linfo, inferred_result)
     end
     # The global cache can only handle objects that codegen understands
-    if !isa(inferred_result, Union{CodeInfo, Vector{UInt8}, ConstAPI})
+    if !isa(inferred_result, Union{CodeInfo, Vector{UInt8}})
```
Comment (aviatesk) [inline base/compiler/typeinfer.jl:375]:
```suggestion
    if !isa(inferred_result, MaybeCompressed)
```

File: base/compiler/typeinfer.jl (hunk 1 above)
Comment (aviatesk) [reply base/compiler/typeinfer.jl:296]: And I think we should also check `is_inlineable_constant(result_type.value)` here.

Comment (staticfloat) [reply base/compiler/typeinfer.jl:296]: > I think we only require is_foldable && is_nothrow here?

I think `is_total()` is defined as ` is_foldable` && `is_nothrow`?

Comment (aviatesk) [reply base/compiler/typeinfer.jl:296]: > I think `is_total()` is defined as `is_foldable` && `is_nothrow`?

Yes, I just remembered it. Perhaps we should have revised its definition as we added the additional effects. "is total" is now a much stronger condition that covers the other guarantees, such as `is_notaskstate` and `is_inaccessiblememonly`. So be safe to use `is_foldable && is_nothrow` here and in the inlining handling.

Comment (staticfloat) [reply base/compiler/typeinfer.jl]: So you're saying that `is_total()` may change its definition in the future?

**Comment by staticfloat** [comment]
I've added a change to try and fix the `.cov` test failure, and while it improves the IR, we're still not getting a `.cov` file.

**Comment by staticfloat** [comment]
I believe I have fixed the last bit of tests related to code coverage.

Code coverage is implemented by inserting a special code coverage IR node into the IR, however it is inserted at optimization time, which is after inference is finished assigning effects to return types.  To fix this, we moved the code coverage check (including scoped coverage selection) into inference time, mark each `InferenceFrame` as needing a coverage node insertion or not, and then taint the effects as "not effect free".  This correctly prevents the coverage nodes from being dropped by later optimization passes, and brings back the coverage we all know and love.

**Comment by staticfloat** [comment]
One remaining question is the second constructor to `OptimizationState`; it does not contain a reference to its `InferenceState`, so we can't receive the `insert_coverage` boolean; because the core compiler does not use this constructor, I assume it's okay to just hardcode `insert_coverage` as `false` here; but if not, I'd appreciate guidance on what to do here instead.

**Pull request merged**